### ISSUE TITLE
add lesson difficulty indicators to lessons page

### ISSUE THREAD

**Issue by fredgibbs** [programminghistorian/jekyll#91]
this is relatively easy to do (if a bit time consuming; but could be divided up among many people), and could be a big help in helping new visitors understand the variety of lessons we have. if such a visitor just happened to skim an advanced lesson but thought it was for beginners, that may be the last thing they view on the site. we'd have to decide on a scale; i'd suggest something like 1-5 to start with.

**Comment by acrymble** [comment]
I do support this idea. Is 5 too many gradients? What's the difference between a 3 and a 4?
I do think there are lessons a new person can jump right into and achieve with no other background, and others that require some pre-work or pre-skills. What does everyone think of just beginner and intermediate as a starting point?

**Comment by fredgibbs** [comment]
we definitely have a few advanced lessons (i'd say), so i'd say we need at
least 3 levels--but i still think that's too few. i would say the
difference b/w a 3 and 4 is that a 3 is a thoroughly intermediate lesson
(compared to other lessons) with one or two advanced concepts integral to
the lesson (that other "3" lessons don't have). perhaps we could try to
rank the lessons 1-5 and if we have virtually no 2s or 4s then we could get
rid of them for simplicity. i'd be surprised if that were the case, but
then again i haven't systematically thought through all the lessons.
overall, i hope we end up with something like a bell curve across all
lessons, which is why i think 5 categories would be helpful for us to use
when helping authors conceive of or write lessons.

On Sun, May 10, 2015 at 8:32 AM, Adam Crymble <notifications@github.com>
wrote:

> I do support this idea. Is 5 too many gradients? What's the difference
> between a 3 and a 4?
> I do think there are lessons a new person can jump right into and achieve
> with no other background, and others that require some pre-work or
> pre-skills. What does everyone think of just beginner and intermediate as a
> starting point?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/programminghistorian/jekyll/issues/91#issuecomment-100647310>
> .
>



-- 
frederick w gibbs | assistant professor of history | univ. of new mexico
fredgibbs.net | @fredgibbs
<http://www.twitter.com/fredgibbs>

**Comment by acrymble** [comment]
What if we let readers decide?

**Comment by ianmilligan1** [comment]
I think this is a great idea, too!

FWIW, I worry that five gradients might be a bit too much, whereas three is perhaps a sweet spot (Fred's idea of going through and seeing might make this more informed - perhaps there is too much complexity for three categories to adequately capture..).

Something like:
- Beginner: The sort of lesson you could pick up with no prior knowledge (i.e. command line with bash);
- Intermediate: Maybe lessons that require some prior knowledge or a previous *Programming Historian* lesson, i.e. (wget)
- Advanced: Lessons that require both prior knowledge/previous lesson as well as some more sophisticated concepts, or that will take a lot of time to painstakingly go through an understand (i.e. the classifier lesson)

I think this works well. DHSI has three gradients which I think works well, in concept at least.

**Comment by ahegel** [comment]
In our Skype call today, we agreed to go through our previously published lessons and mark their difficulty (enter it at the end of each line after the colon). We're starting with "Beginner," "Intermediate," and "Advanced," but if, as you're going through the lessons, you feel that we need more categories, leave a comment!

FG: i've used 1=beginner; 2=intermediate; 3=advanced. i'm thinking we could have 1-3 little computer icons to indicate difficulty (i'm sure i'm unconsciously thinking of menus with little chili peppers to indicate spiciness). 
- [ ] Intro to the Zotero API - Amanda Morton (Fred): 1
- [ ] Creating New Items in Zotero - Amanda Morton (Fred): 1
- [ ] Counting Frequencies from Zotero Items - Spencer Roberts (Fred): 1
- [ ] Preserving Research Data - James Baker (Bill/Adam): 1
- [ ] Sustainable Authorship in Plain Text using Pandoc and Markdown - Dennis Tenen and Grant Wythoff (Fred): 2
- [ ] Introduction to the Bash Command Line - Ian Milligan and James Baker (Allison/Adam): 1
- [ ] Counting and mining research data with Unix - James Baker and Ian Milligan (Allison/Adam): 2
- [ ] Cleaning Data with OpenRefine - Seth van Hooland, Ruben Verborgh, Max De Wilde (Adam): 2
- [ ] Understanding Regular Expressions - Doug Knox (Adam): 2
- [ ] Cleaning OCR’d Text with Regular Expressions - Laura Turner O’Hara (Fred): 2
- [ ] Transliterating non-ASCII Characters with Python - Seth Bernstein (Adam): 2
- [ ] Generating an Ordered Data Set from an OCR Text File - Jon Crump (Fred): 3
- [ ] Corpus Analysis with Antconc - Healther Froehlich (Fred): 1
- [ ] Supervised Classification with a Naive Bayesian - Vilja Hulden (Bill/Adam): 3
- [ ] Getting Started with Topic Modeling and MALLET - Shawn Graham, Scott Weingart, & Ian Milligan (Adam): 2
- [ ] Intro to Google Maps and Google Earth - Jim Clifford, Josh MacFadyen, Daniel Macfarlane (Adam): 1
- [ ] Installing QGIS 2.0 and Adding Layers - Jim Clifford, Josh MacFadyen, Daniel Macfarlane (Adam): 1
- [ ] Creating New Vector Layers in QGIS 2.0 - Jim Clifford, Josh MacFadyen, Daniel Macfarlane (Adam): 2
- [ ] Georeferencing in QGIS 2.0 - Jim Clifford, Josh MacFadyen, Daniel Macfarlane (Adam): 2
- [ ] Creating Network Diagrams from Historical Sources - Marten Düring (Fred): 1
- [ ] Up and Running with Omeka.net - Miriam Posner (Adam): 1
- [ ] Creating an Omeka.net Exhibit - Miriam Posner (Adam): 2
- [ ] Datamining the Internet Archive Collection - Caleb McDaniel (Bill/Adam): 
- [ ] Automated Downloading with Wget - Ian Milligan (Adam): 1
- [ ] Applied Archival Downloading with Wget - Kellen Kurschinski (Ian): 2
- [ ] Intro to Beautiful Soup - Jeri Wieringa: 2
- [ ] Downloading Multiple Records Using Query Strings - Adam Crymble (Fred): 2

**Comment by ianmilligan1** [comment]
Just went through and I think this looks good - in my mind a Difficulty One is something that can be done as a completely self-contained lesson, whereas the others are more advanced or have dependencies (i.e. "Applied Archival Downloading with Wget" really requires the introductory lesson). 

The only one I might change would be "Sustainable Authorship in Plain Text using Pandoc and Markdown" - it predates our CLI lessons, and I think it does a really good job of laying things out clearly (down to reminding us where to find the terminal applications on multiple platforms).

**Comment by acrymble** [comment]
I like the idea of defined criteria, Ian. Can we go down that road and see if we need to re-think?

**Comment by acrymble** [comment]
It looks like this got lost, but we've done all of the work. Just need to add to YAML I presume?

**Comment by wcaleb** [comment]
I've added the difficulty indicators to the lessons in the checklist above. From now on, when adding a new lesson, we need to add a `difficulty: 1` (or `2` or `3`) line to the YAML metadata for the lesson. There are some lessons currently that still need that added---I only did the ones listed in the checklist.

Two other things still need to be done before the issue is closed:
1. Somewhere on the site we should describe for readers what the levels mean, and perhaps write a Javascript so that hovering over the difficulty level on a given lesson shows a pop-up definition.
2. Update either our editorial instructions or guidelines for reviewers to include a section about deciding what difficulty level to give to a lesson.
